xdsl/transforms/function_transformations.py

```diff
@@ -0,0 +1,37 @@
+from xdsl.context import MLContext
+from xdsl.dialects import builtin, func
+from xdsl.dialects.builtin import ArrayAttr, DictionaryAttr, StringAttr
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+
+
+class ArgNamesToArgAttrsPass(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, op: func.FuncOp, rewriter: PatternRewriter, /):
+        arg_attrs = list(op.arg_attrs or (len(op.args) * [DictionaryAttr({})]))
+        new_arg_attrs: list[DictionaryAttr] = []
+        for arg, arg_attr in zip(op.args, arg_attrs):
+            if arg.name_hint:
+                d_attr = DictionaryAttr(
+                    {"llvm.name": StringAttr(arg.name_hint), **arg_attr.data}
+                )
+            else:
+                d_attr = arg_attr
+            new_arg_attrs.append(d_attr)
+
+        op.arg_attrs = ArrayAttr(new_arg_attrs)
+        rewriter.has_done_action = True
```

Comment:
- [ ] It feels like the first thing to do would be to check whether any of the args aren't anonymous, and to bail out early if that's the case. Could you add a test case for this, also?
- [ ] could you also check whether the old attrs are the same as the new attrs, and avoid has_done_action if they are the same?
- [ ] nit: it would be nice to refactor this to a tuple comprehension instead of appending to a list, just for style reasons

Reply:
I had it as a tuple-comp, and after refactoring it am glad that is encouraged 😃

Reply:
I'm a fan :)